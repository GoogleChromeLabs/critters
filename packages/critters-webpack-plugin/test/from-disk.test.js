/**
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

import { compile, compileToHtml, parseDom, readFile } from './_helpers.js';

function configureEmit(config) {
  config.module.rules.push(
    {
      test: /\.html$/,
      loader: 'file-loader?name=[name].[ext]'
    },
    {
      test: /\.css$/,
      loader: 'file-loader?name=[name].[ext]&emitFile=true'
    }
  );
}

function configureNoEmit(config) {
  config.module.rules.push(
    {
      test: /\.html$/,
      loader: 'file-loader?name=[name]-fd.[ext]'
    },
    {
      test: /\.css$/,
      loader: 'file-loader?name=[name].[ext]&emitFile=false'
    }
  );
}

test('webpack compilation', async () => {
  const info = await compile('fixtures/fromDisk/index.js', configureEmit);
  expect(info.assets).toHaveLength(3);
  const html = await readFile('fixtures/fromDisk/index.html');
  expect(html).toMatchSnapshot();
  expect(html).toMatch(/link\srel="stylesheet"\shref=".*?"/);
});

describe('Usage without comp.assets with loading from disk',() => {
  let html, document;
  beforeAll(async () => {
    await compileToHtml('fromDisk', configureNoEmit);
    html = await readFile('fixtures/fromDisk/dist/index-fd.html');
    document = parseDom(html);
  });
  it('should process the first html asset',() => {
    expect(html).toMatchSnapshot();
    expect(html).toMatch(/link\srel="preload".*?as="style"/);
    expect(document.querySelectorAll('style')).toHaveLength(1);
  });
});

describe('Usage with comp.assets',() => {
  let output;
  beforeAll(async () => {
    output = await compileToHtml('fromDisk', configureEmit);
  });
  it('should process the first html asset', () => {
    const { html, document } = output;
    expect(html).toMatchSnapshot();
    expect(html).toMatch(/link\srel="preload".*?as="style"/);
    expect(document.querySelectorAll('style')).toHaveLength(1);
  });
});
