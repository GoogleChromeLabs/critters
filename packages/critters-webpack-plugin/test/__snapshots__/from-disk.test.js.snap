// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Usage with comp.assets and pruneSource should process the first html asset 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
    <title>Loading asset from disk</title>
    <style>:root {
  font-size: 10px;
}
html, body {
  width: 100%;
  height: 100%;
}
</style><link rel=\\"preload\\" href=\\"style.css\\" as=\\"style\\">
  </head>
  <body>
    <h1>Some HTML Here</h1>
  

<link rel=\\"stylesheet\\" href=\\"style.css\\"></body></html>"
`;

exports[`Usage with comp.assets should process the first html asset 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
    <title>Loading asset from disk</title>
    <style>:root {
  font-size: 10px;
}
html, body {
  width: 100%;
  height: 100%;
}
</style><link rel=\\"preload\\" href=\\"style.css\\" as=\\"style\\">
  </head>
  <body>
    <h1>Some HTML Here</h1>
  

<link rel=\\"stylesheet\\" href=\\"style.css\\"></body></html>"
`;

exports[`Usage without comp.assets with loading from disk should process the first html asset 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
    <title>Loading asset from disk</title>
    <style>:root {
  font-size: 10px;
}
html, body {
  width: 100%;
  height: 100%;
}
</style><link rel=\\"preload\\" href=\\"style.css\\" as=\\"style\\">
  </head>
  <body>
    <h1>Some HTML Here</h1>
  

<link rel=\\"stylesheet\\" href=\\"style.css\\"></body></html>"
`;

exports[`webpack compilation 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <title>Loading asset from disk</title>
    <link rel=\\"stylesheet\\" href=\\"style.css\\">
  </head>
  <body>
    <h1>Some HTML Here</h1>
  </body>
</html>
"
`;
